cli/src/util.rs

```diff
@@ -36,65 +38,20 @@ pub fn decode_psbt_file(
     Ok(psbt)
 }
 
-use bitcoin::secp256k1::{All, Secp256k1};
-use bitcoin::util::bip32::ExtendedPrivKey;
-use bitcoin::util::sighash::Prevouts;
-use bitcoin::util::taproot::TapLeafHash;
-use bitcoin::util::taproot::TapSighashHash;
-use bitcoin::SchnorrSig;
-
-use bitcoin::TxOut;
-use bitcoin::XOnlyPublicKey;
-pub fn sign_psbt(
-    xpriv: &ExtendedPrivKey,
-    mut b: PartiallySignedTransaction,
-    secp: &Secp256k1<All>,
-) -> Result<PartiallySignedTransaction, std::io::Error> {
-    let tx = b.clone().extract_tx();
-    let utxos: Vec<TxOut> = b
-        .inputs
-        .iter()
-        .map(|o| o.witness_utxo.clone())
-        .collect::<Option<Vec<TxOut>>>()
-        .ok_or_else(|| input_err("Could not find one of the UTXOs to be signed over"))?;
-    let untweaked = xpriv.to_keypair(secp);
-    let pk = XOnlyPublicKey::from_keypair(&untweaked);
-    let mut sighash = bitcoin::util::sighash::SighashCache::new(&tx);
-    let input_zero = &mut b.inputs[0];
-    use bitcoin::schnorr::TapTweak;
-    let tweaked = untweaked
-        .tap_tweak(secp, input_zero.tap_merkle_root)
-        .into_inner();
-    let _tweaked_pk = tweaked.public_key();
-    let hash_ty = bitcoin::util::sighash::SchnorrSighashType::All;
-    let prevouts = &Prevouts::All(&utxos);
-    let mut get_sig = |path, kp| {
-        let annex = None;
-        let sighash: TapSighashHash = sighash
-            .taproot_signature_hash(0, prevouts, annex, path, hash_ty)
-            .expect("Signature hash cannot fail...");
-        let msg =
-            bitcoin::secp256k1::Message::from_slice(&sighash[..]).expect("Size must be correct.");
-        let sig = secp.sign_schnorr_no_aux_rand(&msg, kp);
-        SchnorrSig { sig, hash_ty }
-    };
-    if input_zero.tap_internal_key == Some(pk.0) {
-        let sig = get_sig(None, &tweaked);
-        input_zero.tap_key_sig = Some(sig);
-    }
-    for tlh in input_zero
-        .tap_scripts
-        .values()
-        .map(|(script, ver)| TapLeafHash::from_script(script, *ver))
-    {
-        let sig = get_sig(Some((tlh, 0xffffffff)), &untweaked);
-        input_zero.tap_script_sigs.insert((pk.0.clone(), tlh), sig);
-    }
-    Ok(b)
-}
-
-fn input_err(s: &str) -> std::io::Error {
-    std::io::Error::new(std::io::ErrorKind::InvalidInput, s)
+/// Reads a PSBT either from a string or from stdin
+pub async fn get_psbt_from(
+    psbt_str: Option<&str>,
+) -> Result<PartiallySignedTransaction, Box<dyn Error>> {
```

Comment:
If you generalize over reader/asyncreader here you can obviate the need for Option, making the API more consistent.

Reply:
not sure what that means... the option comes from the fact that the arg parser in cli may or may not have a psbt_str?

(this part is also mostly a moveonly refactoring btw)

Reply:
My point here is that you should be able to trivially construct a reader of &str, and you can also can create one for stdin. This unifies the input type under an interface that describes precisely the way you are using it. Right now, the `None` side of the option is "magic". Granted it's documented in the comments, but by generalizing over reader, you could pull a PSBT from a network call, a file, etc. and the caller should be able to trivially construct readers for these things.

Reply:
Also noted on the move only refactor, there's a reason I didn't "request changes" but rather just provided commentary.